Add debug tool to backfill order fulfillment status

Nikschavan opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).

### Changes proposed in this Pull Request:

Part of WOOPLUG-5707.

Depends on https://github.com/woocommerce/woocommerce/pull/61487

This PR adds a debug tool to backfill the `fulfillment_status` column in the `wc_order_stats` table for existing orders. This is part of the fulfillments feature implementation.

**Key Changes:**

1. **DataStore.php** - Added `add_fulfillment_status_column()` method to add the column if it doesn't exist
2. **Analytics.php** - Added regeneration tool with Action Scheduler-based batch processing:
   - Processes orders in batches of 100
   - Reads `_fulfillment_status` order meta
   - Updates `wc_order_stats.fulfillment_status` column
   - Includes progress tracking and error handling
   - Only shows tool when fulfillments feature is enabled
3. **Tests** - Added comprehensive test to verify the migration works correctly

**Technical Details:**

- Uses Action Scheduler for background processing to prevent timeouts
- Batch size: 100 orders per action
- Progress tracking via transients
- Race condition protection
- Consistent transient cleanup on completion

Builds on: #new-column-wc-order-stats

### How to test the changes in this Pull Request:

1. Switch to this PR branch
2. Ensure the `fulfillment_status` column does NOT exist on the `wc_order_stats` table (drop it if it exists):
   ```sql
   ALTER TABLE wp_wc_order_stats DROP COLUMN fulfillment_status;
   ```
3. Enable the fulfillments feature if not already enabled
4. Create some test orders with various fulfillment statuses
5. Navigate to WooCommerce > Status > Tools
6. Find the "Regenerate order fulfillment status" tool and click "Regenerate"
7. Confirm Action Scheduler actions are scheduled and executed:
   - Go to WooCommerce > Status > Scheduled Actions
   - Look for actions with hook `woocommerce_analytics_regenerate_order_fulfillment_status`
8. Verify the column exists and has order fulfillment statuses populated:
   ```sql
   SELECT order_id, fulfillment_status FROM wp_wc_order_stats LIMIT 10;
   ```
9. Run the PHPUnit test to verify functionality:
   ```bash
   pnpm test:php:env -- --filter test_regenerate_order_fulfillment_status_updates_orders_with_fulfillments
   ```

### Testing that has already taken place:

- Unit tests added to verify the migration correctly updates orders with fulfillments
- Verified Action Scheduler batch processing works correctly
- Confirmed transient cleanup and race condition protection

### Changelog entry

- [ ] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

- [x] Patch

#### Type

- [x] Add - Adds functionality

#### Message

Add debug tool to backfill order fulfillment status in analytics table

</details>

Nikschavan committed:
Add fulfillment_status column to wc_order_stats table for new installations

Nikschavan committed:
Add tests to verify changes to table schema based on if it's a fresh install

Nikschavan committed:
cs fixes

Nikschavan committed:
cs fixes

Nikschavan committed:
cs fixes

Nikschavan committed:
dont cache status of column

Nikschavan committed:
Switch DataStore fulfillment column check to use option instead of static variable

The static variable cache was breaking tests. Now using option-based
approach for checking if fulfillment_status column exists.

Nikschavan committed:
Revert "Switch DataStore fulfillment column check to use option instead of static variable"

This reverts commit 4d14b1066606810dba89d5dcdcd417e5d857c5bd.

Nikschavan committed:
Add Action Scheduler-based order fulfillment status regeneration tool

This commit introduces a background processing system for regenerating order fulfillment status data in Analytics using Action Scheduler.

Changes:
- Add `add_fulfillment_status_column()` method to create fulfillment_status column with proper error handling
- Implement batch processing via Action Scheduler (100 orders per batch)
- Add real-time progress tracking using transients
- Update debug tool UI to show progress and disable button during execution
- Tool automatically hides after completion

Benefits:
- Prevents timeout issues on stores with many orders
- Provides user feedback with "Processed X of Y orders" message
- Prevents duplicate runs with scheduling checks
- Processes in background without blocking admin UI

Nikschavan committed:
Add Action Scheduler-based order fulfillment status regeneration tool

This commit introduces a background processing system for regenerating order fulfillment status data in Analytics using Action Scheduler.

Changes:
- Add `add_fulfillment_status_column()` method to create fulfillment_status column with proper error handling
- Implement batch processing via Action Scheduler (100 orders per batch)
- Add real-time progress tracking using transients
- Update debug tool UI to show progress and disable button during execution
- Tool automatically hides after completion

Benefits:
- Prevents timeout issues on stores with many orders
- Provides user feedback with "Processed X of Y orders" message
- Prevents duplicate runs with scheduling checks
- Processes in background without blocking admin UI

Nikschavan committed:
Don't show the tool if fulfillments feature is disabled

Nikschavan committed:
Don't show the tool if fulfillments feature is disabled

Nikschavan committed:
Keep completed status consistent

Nikschavan committed:
Keep completed status consistent

Nikschavan committed:
Add test to confirm backfill of order fulfillment status synced to order stats table

Nikschavan committed:
Add test to confirm backfill of order fulfillment status synced to order stats table

Nikschavan committed:
cs fixes

Nikschavan committed:
cs fixes

Nikschavan committed:
Reapply "Switch DataStore fulfillment column check to use option instead of static variable"

This reverts commit 857226dffe7b6e404eef643226a00215049b6461.

Nikschavan committed:
Regenerate the table status option if it does not exist

Nikschavan committed:
Delete option to force regenerate status before tests

Nikschavan committed:
Delete option instead

Nikschavan committed:
Don't set option for table column exists

The option is set by has_fulfillment_status_column if the option does not exist

Nikschavan committed:
Fix incorrect phpdoc param types

Nikschavan committed:
cs fixes

Nikschavan committed:
cs fixes

Nikschavan committed:
cs fixes

Nikschavan committed:
cs fixes

Nikschavan committed:
Cs fixes

Nikschavan committed:
Cs fixes

Nikschavan committed:
CS fixes

Nikschavan committed:
Set flag to indicate the table has fulfillment column

Nikschavan committed:
Fix inline comments

Nikschavan committed:
Merge branch 'debug-tool-backfill-order-fulfillment-status' of https://github.com/woocommerce/woocommerce into debug-tool-backfill-order-fulfillment-status

Nikschavan commented on plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/reports/class-wc-tests-reports-orders-stats.php:
```diff
@@ -6502,4 +6504,133 @@ public function test_registered_returning_customer() {
+		$this->add_fulfillment_to_order( $order_1, 'fulfilled', $product );
+		$this->add_fulfillment_to_order( $order_2, 'partially_fulfilled', $product );
+		$this->add_fulfillment_to_order( $order_3, 'unfulfilled', $product );
+		// Orders 3 and 4 have no fulfillments.
```
Fixed

Nikschavan commented on plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/reports/class-wc-tests-reports-orders-stats.php:
```diff
@@ -6502,4 +6504,133 @@ public function test_registered_returning_customer() {
+
+		WC_Helper_Queue::run_all_pending( 'wc-admin_import_orders' );
+
+		// Add fulfillments for only orders 0, 1, 2.
```
Fixed

Nikschavan committed:
Set option also if the column does not exist to avoid repeat calls to database

Nikschavan committed:
cs fixes

Nikschavan committed:
Include the column in schema if it already exists

Nikschavan committed:
Add check if the fulfillments feature is enabled

Nikschavan committed:
Squashed commit of the following:

commit c98efe845dd27ffd528068dda8c0272875be82d5
Author: Nikhil <nikhil.chavan@automattic.com>
Date:   Tue Nov 4 11:39:20 2025 +0530

    Add check if the fulfillments feature is enabled

commit 6fcca417754b260e8cd407d90c66bfab19b6d909
Author: Nikhil <nikhil.chavan@automattic.com>
Date:   Tue Nov 4 11:37:07 2025 +0530

    Include the column in schema if it already exists

commit 20227cb78786c2684fe7a61ee99a30c4e07b06a1
Author: Nikhil <nikhil.chavan@automattic.com>
Date:   Mon Nov 3 15:07:20 2025 +0530

    Set option also if the column does not exist to avoid repeat calls to database

Nikschavan committed:
Merge branch 'new-column-wc-order-stats' into debug-tool-backfill-order-fulfillment-status

Nikschavan committed:
Merge branch 'trunk' into new-column-wc-order-stats

Nikschavan committed:
Update test to test with or without column

Nikschavan committed:
use correct option values

Nikschavan committed:
Merge branch 'new-column-wc-order-stats' into debug-tool-backfill-order-fulfillment-status

Nikschavan committed:
Merge branch 'trunk' into debug-tool-backfill-order-fulfillment-status

Nikschavan committed:
Merge branch 'trunk' into debug-tool-backfill-order-fulfillment-status

Nikschavan committed:
Verify the structure of the progress array before using array keys

Nikschavan committed:
Validate the page attribute

Nikschavan committed:
Log the error if progress transient expired

Nikschavan committed:
Increase the migration counter even if current order is not valid

Nikschavan committed:
Increase the counter even if the current order failed to update

Log error in case of error

Nikschavan committed:
Properly cleanup after the test

Nikschavan commented on plugins/woocommerce/src/Internal/Admin/Analytics.php:
```diff
@@ -175,6 +186,199 @@ public function register_cache_clear_tool( $debug_tools ) {
+			if ( false !== $result ) {
+				++$updated_count;
+			}
+		}
```
This is addressed in 3bec6c68e9bc883bae0d435cc55364b76a8ea1de and cf1fe31a051cc3ebd1a8e00d250e427e1c54f065

Nikschavan committed:
Guard against unexpected transient payloads before reading status

Nikschavan committed:
Always cleanup

Nikschavan committed:
Create the column if the fulfillment feature is enabled

Nikschavan committed:
Update tests to account for feature to be enabled and disabled

Nikschavan commented on plugins/woocommerce/includes/class-wc-install.php:
```diff
@@ -1244,7 +1245,7 @@ private static function should_enable_hpos_for_new_shop() {
 		global $wpdb;
 
-		$should_have_fulfillment_column = self::is_new_install();
+		$should_have_fulfillment_column = self::is_new_install() && FeaturesUtil::feature_is_enabled( 'fulfillments' );
```
@louwie17 - Addressed feedback from https://github.com/woocommerce/woocommerce/pull/61487#discussion_r2491245185

Nikschavan committed:
Cleanup correct filter name

Nikschavan committed:
Clean fulfillments

Nikschavan committed:
Update test comment

Nikschavan committed:
Check if the table exists before checking if the column exists

Nikschavan committed:
cs fixes

Nikschavan committed:
cs fixes

Nikschavan committed:
Add changelog entry

Nikschavan committed:
cs fixes

Nikschavan committed:
Unschedule all scheduled jobs from test

Nikschavan committed:
Update comment to match assertion

Nikschavan commented on plugins/woocommerce/src/Internal/Admin/Analytics.php:
```diff
@@ -175,6 +186,206 @@ public function register_cache_clear_tool( $debug_tools ) {
+	 *
+	 * @return string Success message or error message.
+	 */
+	public function run_regenerate_order_fulfillment_status_tool() {
```
@louwie17 - I was curious about thoughts here. I chose to use Action Scheduler to process the backfill as I thought that this is a tool that can be used much later when a site has a lot more fulfillments. The action scheduler method will work for any size of updates required. That said, I see there are [similar migrations](https://github.com/woocommerce/woocommerce/blob/fe4b89e990be3b512febba5ba32b295073653476/plugins/woocommerce/src/Internal/CostOfGoodsSold/CostOfGoodsSoldController.php#L113-L118) that do it with a simple `UPDATE` query.

Downside of action schduler approahc is the code is bit more complex than the simple update query so I was curious if you have any preferance/thoughts on this

louwie17 replied:
Yeah at the moment given the fulfillment feature is still brand new I don't have a strong preference and I would say that skipping the Action Scheduler is totally fine and doing the update directly would work well.

> That said, I see there are [similar migrations](https://github.com/woocommerce/woocommerce/blob/fe4b89e990be3b512febba5ba32b295073653476/plugins/woocommerce/src/Internal/CostOfGoodsSold/CostOfGoodsSoldController.php#L113-L118) that do it with a simple UPDATE query.

This example is interesting, because it seems like one scenario where maybe the action scheduler would be of benefit for stores with a huge amount of products. But I suppose a large set of products is less likely over a large set of orders ( over time ).

louwie17 replied:
In relation to this, currently the fulfillment feature cannot be explicitly enabled by a merchant, but I assume it will be an option as part of the **Settings > Advanced > Features** at some point.
In which case, we may want to auto trigger this update script when the feature is explicitly enabled. This would also mean that the `fulfillment` table wouldn't contain any rows yet, so it would only be adding the new column.

Nikschavan replied:
> Yeah at the moment given the fulfillment feature is still brand new

What I was thinking is, this tool will stay in WC core, and I imagine there will be stores that discover this tool much after they start using fulfillments.

Nikschavan replied:
For now ill use direct SQL query, IMO we can refactor this to use action scheduler in the future if required

louwie17 commented on plugins/woocommerce/src/Admin/API/Reports/Orders/Stats/DataStore.php:
```diff
@@ -692,7 +692,22 @@ public static function has_fulfillment_status_column() {
+		$table_name = self::get_db_table_name();
+
+		// Check if the table exists.
+		$table_exists = $wpdb->get_var(
+			$wpdb->prepare(
+				// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared -- Table name cannot be prepared.
+				'SHOW TABLES LIKE %s',
+				$table_name
+			)
+		);
```
There is a situation where this is called and the table doesn't exist?

Nikschavan replied:
Due to [feature exists check](https://github.com/woocommerce/woocommerce/pull/61721/files#r2493193877), when creating the tables, it can go into the if condition below to check if the column exists, this is when the table does not exist. This was caught by many unit tests failing because after the above condition was added

louwie17 commented on plugins/woocommerce/src/Internal/Admin/Analytics.php:
```diff
@@ -175,6 +186,206 @@ public function register_cache_clear_tool( $debug_tools ) {
+				array( 'order_id' => $order_id ),
+				array( '%s' ),
+				array( '%d' )
+			);
```
Not a 100% sure if it saves a lot of query time, but could we do a bulk update for each status?
Meaning in one query we update all the order ids that are 'fullfilled', etc...

louwie17 commented on plugins/woocommerce/src/Internal/Admin/Analytics.php:
```diff
@@ -175,6 +186,206 @@ public function register_cache_clear_tool( $debug_tools ) {
+				$per_page,
+				$offset
+			)
+		);
```
Instead of using offset Cursor suggested to use a "cursor" based approach, by using the order id as the cursor:
```
// Instead of OFFSET, use WHERE clause
$last_order_id = get_transient( 'woocommerce_analytics_fulfillment_last_order_id' ) ?: 0;

$order_ids = $wpdb->get_col(
    $wpdb->prepare(
        "SELECT DISTINCT entity_id FROM {$fulfillments_table} 
        WHERE entity_type = 'WC_Order' 
        AND date_deleted IS NULL 
        AND entity_id > %d 
        ORDER BY entity_id ASC 
        LIMIT %d",
        $last_order_id,
        $per_page
    )
);

// After processing, store the last ID
if ( ! empty( $order_ids ) ) {
    set_transient( 'woocommerce_analytics_fulfillment_last_order_id', end( $order_ids ), DAY_IN_SECONDS );
}

// Schedule next batch - same logic
if ( count( $order_ids ) === $per_page ) {
    // Schedule next batch
}
```

louwie17 reviewed (changes requested):
Thanks @Nikschavan, this tested well! I did leave a couple suggestions within the code, and do agree with you that using the action scheduler for this is probably overkill and we can just trigger an update directly.

Nikschavan committed:
Merge branch 'trunk' into debug-tool-backfill-order-fulfillment-status

Nikschavan committed:
Refactor order fulfillment status migration to use direct SQL query.

Replace Action Scheduler background processing with a single SQL query that directly copies _fulfillment_status from order meta to the analytics table. This simplifies the migration from ~150 lines to ~30 lines and provides instant execution instead of background batching.

Nikschavan committed:
run the tool directly as we no longer are running action scheduler

Nikschavan committed:
Ignore table name prepare

Nikschavan committed:
cs fixes

Nikschavan committed:
If HOPS is disabled, use post meta instead of order meta

Nikschavan committed:
Merge branch 'trunk' into debug-tool-backfill-order-fulfillment-status

Nikschavan committed:
cs fixes

Nikschavan committed:
Merge branch 'trunk' into debug-tool-backfill-order-fulfillment-status

louwie17 commented on plugins/woocommerce/src/Internal/Admin/Analytics.php:
```diff
@@ -175,6 +180,97 @@ public function register_cache_clear_tool( $debug_tools ) {
+		}
+
+		// phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching
+		$updated = $wpdb->query(
+			$wpdb->prepare(
+				// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared -- Table and column names cannot be prepared.
+				"UPDATE {$order_stats_table} os INNER JOIN {$order_meta_table} om ON os.order_id = om.{$order_meta_column}
+				SET os.fulfillment_status = CASE
+					WHEN om.meta_value = %s THEN NULL
+					ELSE om.meta_value
+				END
+				WHERE om.meta_key = %s",
+				'no_fulfillments',
+				'_fulfillment_status'
+			)
+		);
```
Nice!!

louwie17 reviewed (approved):
LGTM 🚀 Thanks for the updates, this tested well, nice work!!

Nikschavan committed:
Merge branch 'trunk' into debug-tool-backfill-order-fulfillment-status
